### PR TITLE
merge bitcoin#13311: Don't edit Chainparams after initialization

### PR DESCRIPTION
**Pull request by rkarthik2k21**
Backport #13311: Don't edit Chainparams after initialization

### CODE REVIEW

File: src/chainparams.cpp [hunk 1]
```diff
@@ -807,15 +742,15 @@ class CDevNetParams : public CChainParams {
         nDefaultPort = 19799;
         nPruneAfterHeight = 1000;
 
+        UpdateDevnetSubsidyAndDiffParametersFromArgs(args);
+
```
Comment (UdjinM6) [inline src/chainparams.cpp:746]:
```suggestion

```

File: src/chainparams.h [hunk 2]
```diff
@@ -152,49 +151,4 @@ const CChainParams &Params();
...
- */
-void UpdateLLMQTestParams(int size, int threshold);
-
-/**
- * Allows modifying parameters of the devnet LLMQ
- */
-void UpdateLLMQDevnetParams(int size, int threshold);
-
-#endif // BITCOIN_CHAINPARAMS_H
+#endif // BITCOIN_CHAINPARAMS_H
```
Comment (UdjinM6) [inline src/chainparams.h:154]:
```suggestion
#endif // BITCOIN_CHAINPARAMS_H

```

File: src/chainparams.cpp [hunk 3]
```diff
@@ -860,6 +814,48 @@ class CDevNetParams : public CChainParams {
             0.01                          // * estimated number of transactions per second
         };
     }
+
+ /**
```
Comment (UdjinM6) [inline src/chainparams.cpp:818]:
```suggestion
    /**
```

File: src/chainparams.cpp [hunk 4]
```diff
@@ -1008,17 +1007,74 @@ class CRegTestParams : public CChainParams {
...
 
     /**
      * Allows modifying the Version Bits regtest parameters.
      */
-    void UpdateVersionBitsParameters(Consensus::DeploymentPos d, int64_t nStartTime, int64_t nTimeout)
+    void UpdateVersionBitsParameters(Consensus::DeploymentPos d, int64_t nStartTime, int64_t nTimeout, int64_t nWindowSize, int64_t nThresholdStart, int64_t nThresholdMin, int64_t nFalloffCoeff)
     {
         consensus.vDeployments[d].nStartTime = nStartTime;
         consensus.vDeployments[d].nTimeout = nTimeout;
+               if (nWindowSize != -1) {
```
Comment (UdjinM6) [inline src/chainparams.cpp:1021]:
```suggestion
        if (nWindowSize != -1) {
```

File: src/chainparams.cpp [hunk 5]
```diff
@@ -1028,22 +1084,42 @@ void CRegTestParams::UpdateVersionBitsParametersFromArgs(const ArgsManager& args
     for (const std::string& strDeployment : args.GetArgs("-vbparams")) {
         std::vector<std::string> vDeploymentParams;
         boost::split(vDeploymentParams, strDeployment, boost::is_any_of(":"));
-        if (vDeploymentParams.size() != 3) {
-            throw std::runtime_error("Version bits parameters malformed, expecting deployment:start:end");
+       if (vDeploymentParams.size() != 3 && vDeploymentParams.size() != 5 && vDeploymentParams.size() != 7) {
```
Comment (UdjinM6) [inline src/chainparams.cpp:1087]:
```suggestion
        if (vDeploymentParams.size() != 3 && vDeploymentParams.size() != 5 && vDeploymentParams.size() != 7) {
```

File: src/chainparamsbase.cpp [hunk 6]
```diff
@@ -18,11 +18,24 @@ const std::string CBaseChainParams::REGTEST = "regtest";
...
+    gArgs.AddArg("-llmqinstantsend=<quorum name>", "Override the default LLMQ type used for InstantSend. Allows using InstantSend with smaller LLMQs. (default: llmq_50_60, devnet-only)", false, OptionsCategory::CHAINPARAMS);
+    gArgs.AddArg("-llmqtestparams=<size>:<threshold>", "Override the default LLMQ size for the LLMQ_TEST quorum (default: 10:6, regtest-only)", true, OptionsCategory::CHAINPARAMS);
+    gArgs.AddArg("-minimumdifficultyblocks=<n>", "The number of blocks that can be mined with the minimum difficulty at the start of a chain (default: 0, devnet-only)", false, OptionsCategory::CHAINPARAMS);
     gArgs.AddArg("-regtest", "Enter regression test mode, which uses a special chain in which blocks can be solved instantly. "
-                                   "This is intended for regression testing tools and app development.", true, OptionsCategory::CHAINPARAMS);
+                                    "This is intended for regression testing tools and app development.", true, OptionsCategory::CHAINPARAMS);
```
Comment (UdjinM6) [inline src/chainparamsbase.cpp:33]:
```suggestion
                                   "This is intended for regression testing tools and app development.", true, OptionsCategory::CHAINPARAMS);
```

File: src/versionbitsinfo.h [hunk 7]
```diff
@@ -10,6 +10,8 @@ struct VBDeploymentInfo {
     const char *name;
     /** Whether GBT clients can safely ignore this rule in simplified usage */
     bool gbt_force;
+      /** Whether to check current MN protocol or not */
```
Comment (UdjinM6) [inline src/versionbitsinfo.h:13]:
```suggestion
    /** Whether to check current MN protocol or not */
```

File: src/chainparams.cpp [hunk 8]
```diff
@@ -1021,76 +1007,303 @@ class CRegTestParams : public CChainParams {
...
+    LogPrintf("Setting DIP3 parameters to activation=%ld, enforcement=%ld\n", nDIP3ActivationHeight, nDIP3EnforcementHeight);
+    UpdateDIP3Parameters(nDIP3ActivationHeight, nDIP3EnforcementHeight);
 }
 
-void SelectParams(const std::string& network)
+void CRegTestParams::UpdateDIP8ParametersFromArgs(const ArgsManager& args)
 {
-    SelectBaseParams(network);
-    globalChainParams = CreateChainParams(network);
+if (!args.IsArgSet("-dip8params")) return;
```
Comment (UdjinM6) [inline src/chainparams.cpp:1155]:
```suggestion
    if (!args.IsArgSet("-dip8params")) return;
```

File: src/chainparams.cpp [hunk 9]
```diff
@@ -1021,76 +1007,303 @@ class CRegTestParams : public CChainParams {
...
+
+std::string strParams = args.GetArg("-dip8params", "");
```
Comment (UdjinM6) [inline src/chainparams.cpp:1157]:
```suggestion
    std::string strParams = args.GetArg("-dip8params", "");
```

File: src/chainparams.cpp [hunk 10]
```diff
@@ -1021,76 +1007,303 @@ class CRegTestParams : public CChainParams {
...
 }
 
-void UpdateLLMQDevnetParams(int size, int threshold)
+void SelectParams(const std::string& network)
 {
-    globalChainParams->UpdateLLMQDevnetParams(size, threshold);
-}
+    SelectBaseParams(network);
+    globalChainParams = CreateChainParams(network);
+}
```
Comment (UdjinM6) [inline src/chainparams.cpp:1309]:
```suggestion
}

```

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

**Review by UdjinM6 — DISMISSED** [review]
LGTM, utACK for squash merge

File: src/chainparams.cpp [hunk 11]
```diff
@@ -705,7 +640,7 @@ class CTestNetParams : public CChainParams {
  */
 class CDevNetParams : public CChainParams {
 public:
-    CDevNetParams(bool fHelpOnly = false) {
+    CDevNetParams(const ArgsManager& args) {
```
Comment (PastaPastaPasta) [inline src/chainparams.cpp:643]: @UdjinM6 should this be marked explicit?

```suggestion
    explicit CDevNetParams(const ArgsManager& args) {
```

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]

File: src/chainparams.cpp (hunk 11 above)
Comment (UdjinM6) [reply src/chainparams.cpp:643]: Good catch!

**Review by UdjinM6 — APPROVED** [review]
re-utACK for squash merge

File: src/test/test_dash.cpp [hunk 12]
```diff
@@ -54,6 +55,8 @@ BasicTestingSetup::BasicTestingSetup(const std::string& chainName)
     InitScriptExecutionCache();
     CCoinJoin::InitStandardDenominations();
     fCheckBlockIndex = true;
+    // Make sure CreateAndProcessBlock() support building <deployment_name> blocks before activating it in these tests.
+    //gArgs.ForceSetArg("-vbparams", strprintf("deployment_name:0:%d", (int64_t)Consensus::BIP9Deployment::NO_TIMEOUT));
```
Comment (PastaPastaPasta) [inline src/test/test_dash.cpp:59]: Why'd we comment this out? Can you add a comment explaining why it was commented out?

File: src/chainparams.cpp [hunk 13]
```diff
@@ -1021,76 +1006,303 @@ class CRegTestParams : public CChainParams {
...
+                throw std::runtime_error(strprintf("Invalid nFalloffCoeff (%s)", vDeploymentParams[6]));
+            }
+        }
+        bool found = false;
+        for (int j=0; j < (int)Consensus::MAX_VERSION_BITS_DEPLOYMENTS; ++j) {
+            if (vDeploymentParams[0] == VersionBitsDeploymentInfo[j].name) {
+                UpdateVersionBitsParameters(Consensus::DeploymentPos(j), nStartTime, nTimeout, nWindowSize, nThresholdStart, nThresholdMin, nFalloffCoeff);
+                found = true;
+                LogPrintf("Setting version bits activation parameters for %s to start=%ld, timeout=%ld, window=%ld, thresholdstart=%ld, thresholdmin=%ld, falloffcoeff=%ld\n",
+                        vDeploymentParams[0], nStartTime, nTimeout, nWindowSize, nThresholdStart, nThresholdMin, nFalloffCoeff);
```
Comment (PastaPastaPasta) [inline src/chainparams.cpp:1121]: Clang format says
```suggestion
                          vDeploymentParams[0], nStartTime, nTimeout, nWindowSize, nThresholdStart, nThresholdMin, nFalloffCoeff);
```

File: src/test/test_dash.cpp (hunk 12 above)
Comment (UdjinM6) [reply src/test/test_dash.cpp:59]: This is simply a placeholder for <deployment_name> deployment that we don't test yet but might want to do so in the future. In the original PR it's used for segwit (which we don't have).

File: src/chainparams.cpp [hunk 14]
```diff
@@ -1021,76 +1006,303 @@ class CRegTestParams : public CChainParams {
...
 {
-    globalChainParams->UpdateLLMQTestParams(size, threshold);
+    if (chain == CBaseChainParams::MAIN)
+        return std::unique_ptr<CChainParams>(new CMainParams());
+    else if (chain == CBaseChainParams::TESTNET)
+        return std::unique_ptr<CChainParams>(new CTestNetParams());
+    else if (chain == CBaseChainParams::DEVNET) {
+        return std::unique_ptr<CChainParams>(new CDevNetParams(gArgs));
+    } else if (chain == CBaseChainParams::REGTEST)
+         return std::unique_ptr<CChainParams>(new CRegTestParams(gArgs));
```
Comment (PastaPastaPasta) [inline src/chainparams.cpp:1299]:
```suggestion
        return std::unique_ptr<CChainParams>(new CRegTestParams(gArgs));
```

File: src/chainparamsbase.cpp [hunk 15]
```diff
@@ -18,10 +18,24 @@ const std::string CBaseChainParams::REGTEST = "regtest";
...
+    gArgs.AddArg("-llmqtestparams=<size>:<threshold>", "Override the default LLMQ size for the LLMQ_TEST quorum (default: 10:6, regtest-only)", true, OptionsCategory::CHAINPARAMS);
+    gArgs.AddArg("-minimumdifficultyblocks=<n>", "The number of blocks that can be mined with the minimum difficulty at the start of a chain (default: 0, devnet-only)", false, OptionsCategory::CHAINPARAMS);
     gArgs.AddArg("-regtest", "Enter regression test mode, which uses a special chain in which blocks can be solved instantly. "
                                    "This is intended for regression testing tools and app development.", true, OptionsCategory::CHAINPARAMS);
     gArgs.AddArg("-testnet", "Use the test chain", false, OptionsCategory::CHAINPARAMS);
+    gArgs.AddArg("-vbparams=<deployment>:<start>:<end>(:<window>:<threshold/thresholdstart>(:<thresholdmin>:<falloffcoeff>))",
+                    "Use given start/end times for specified version bits deployment (regtest-only). "
+                    "Specifying window, threshold/thresholdstart, thresholdmin and falloffcoeff is optional.", true, OptionsCategory::CHAINPARAMS);
```
Comment (PastaPastaPasta) [inline src/chainparamsbase.cpp:37]:
```suggestion
                 "Use given start/end times for specified version bits deployment (regtest-only). "
                 "Specifying window, threshold/thresholdstart, thresholdmin and falloffcoeff is optional.", true, OptionsCategory::CHAINPARAMS);
```

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]
a few clang format suggestions

**Review by PastaPastaPasta — APPROVED** [review]
utACK for squash merge

**Review by UdjinM6 — APPROVED** [review]
re-re-utACK for squash merge :D

**Pull request merged**